src/utils/customActionUtils.ts

```diff
@@ -0,0 +1,185 @@
+        return errorMessage;
+    }
+
+    const config = customActionValidationConfig[targetType];
+    const errors: string[] = [];
+
+    // Validate position
+    if (!arrayIncludesString(config.positions, position)) {
+        errors.push(`Position '${position}' is not supported for ${targetType.toLowerCase()}-level custom actions`);
+    }
+
+    // Validation for Liveboard level custom actions cannot have CONTEXTMENU
+    if (targetType === CustomActionTarget.LIVEBOARD && position === CustomActionsPosition.CONTEXTMENU) {
+        errors.push(`Liveboard-level custom actions cannot have position '${CustomActionsPosition.CONTEXTMENU}'`);
+    }
+
+    // Validation for Spotter level custom actions cannot have PRIMARY
+    if (targetType === CustomActionTarget.SPOTTER && position === CustomActionsPosition.PRIMARY) {
+        errors.push(`Spotter-level custom actions cannot have position '${CustomActionsPosition.PRIMARY}'`);
+    }
+
+    // Check for primary action conflicts
+    if (position === CustomActionsPosition.PRIMARY) {
+        if (primaryActionsPerTarget.has(targetType)) {
+            const errorMessage = `Custom Action Validation: Multiple primary custom actions found for target '${targetType}'. Action '${actionId}' will be ignored`;
+            return errorMessage;
+        }
+        primaryActionsPerTarget.set(targetType, action);
+    }
+
+    // Validate allowed top-level fields
+    Object.keys(action).forEach((key: string) => {
+        if (!arrayIncludesString(config.allowedFields, key)) {
+            errors.push(`Field '${key}' is not supported in ${targetType.toLowerCase()}-level custom actions`);
+        }
+    });
+
+    // Validate metadataIds
+    if (metadataIds && typeof metadataIds === 'object') {
+        Object.keys(metadataIds).forEach((key: string) => {
+            if (!arrayIncludesString(config.allowedMetadataIds, key)) {
+                errors.push(`Field '${key}' in metadataIds is not supported in ${targetType.toLowerCase()}-level custom actions`);
+            }
+        });
+    }
+
+    // Validate dataModelIds
+    if (dataModelIds && typeof dataModelIds === 'object') {
+        Object.keys(dataModelIds).forEach((key: string) => {
+            if (!arrayIncludesString(config.allowedDataModelIds, key)) {
+                errors.push(`Field '${key}' in dataModelIds is not supported in ${targetType.toLowerCase()}-level custom actions`);
+            }
+        });
+    }
+
+    // If there are errors, return error string
+    if (errors.length > 0) {
+        const errorMessage = `Custom Action Validation Error for '${actionId}': ${errors.join('; ')}`;
+        return errorMessage;
+    }
+
+    return null;
+};
+
+/**
+ * Validates and processes custom actions
+ * @param customActions - Array of custom actions to validate
+ * @returns Object containing valid actions and any validation errors
+ */
+export const getCustomActions = (customActions: CustomAction[]): CustomActionsValidationResult => {
```

Comment:
encapsulate all of this into a proper class 

also the return type of this function may mot help much , we should tell customers which object cause which error mayb club them